sympy/simplify/tests/test_simplify.py

```diff
@@ -1046,3 +1049,21 @@ def test_issue_22210():
     d = Symbol('d', integer=True)
     expr = 2*Derivative(sin(x), (x, d))
     assert expr.simplify() == expr
+
+
+def test_issue_23518():
+    M = Symbol("M", positive=True, integer=True)
+    N = Symbol("N", positive=True, integer=True)
+
+    expr = M*N >= N
+    expr = expr.simplify()
+    assert expr == True
```

Comment:
Of the added tests this is the only one that doesn't already pass on master. I'm struggling to understand why there are so many code changes when for this example it could be as simple as this:
```python
In [19]: simplify(expr.rhs - expr.lhs).is_nonnegative
Out[19]: True
```

Reply:
As far as I can see, this test does pass on my local branch. Have there been other changes to the simplification function?

Reply:
I believe it is also more convenient to simplify an expression without needing to check other properties about it manually, so I thought there was value in solving this issue.

Reply:
My point is this: you have changed a lot of code and I'm finding it difficult when I look in the diff to understand precisely why those changes have been made so I looked at the tests. Most of the tests already pass on master without the changes in this PR. This is the only one that this PR actually improves but then I don't understand why the changes in this PR are as complicated as they are in order to accommodate what seems like a fairly simple case.

Reply:
I had introduced `_eval_simplification()` as a helper function because the code inside it used to be in `_eval_simplify()` (both functions in `relational.py`) and it looked very bloated and difficult to understand. I therefore wrote `_eval_simplification()` to split the functionality of `_eval_simplify()` into separate, smaller parts so my solution would be easier to implement and understand.

Reply:
It does not work because of line 469 `m = linear_coeffs(dif, *free)` (I'm not sure how to reference code directly). The problem here is that it tries to get linear coefficients of all free variables (M and N in our example), however, `M*N - N` (the difference) has co-dependent coefficients. So it stops computation and returns the expression as-is. Do you believe I could move my code into `_eval_simplification` as another try-except block underneath this one?

Reply:
We should avoid using `try/except` but I would have thought this logic should somehow be handled in the block that treats `dif` as a polynomial.

Reply:
Where is this block?

Reply:
Oh, I see that block only happens if there is a single free symbol which isn't the case here.

Let me put this another way: the code that you are editing is already quite convoluted and complicated and your changes here seem to add more complication on top. Actually I suspect that the best approach is to somehow simplify the logic in the existing code rather than just adding more on top.

Reply:
I don't think I added any complications on top. I literally allowed the old algorithm to run (which I didn't change), then my new simple algorithm to run. I return the expression that is better out of the two.

Reply:
There are many different kinds of complexity that can be considered:

- Computational complexity - having two algorithms potentially doubles the computational cost.
- Code complexity - having multiple branches makes reasoning about the code more complicated especially when try/except is used because that adds many implicit branches.
- Functional complexity - if there are different branches that apply in slightly different cases then how does that interact with the different kinds of inputs? One branch is for Expr only and the other is for NumberKind only: why do they have different conditions?
- etc

It is almost always better to try to improve the existing code rather than to just add more code on top. The thing to do is study the existing code and consider:

- What is it trying to do?
- In what cases does it work as intended?
- Why does it not work as intended in the case that you are interested in?
- Is there a simpler way of handling all cases of interest?

When I look at the code in the `len(free) >= 2` block I can see that it only attempts to handle linear equations. But I can also see that while it's quite complicated it is mostly just doing something quite simple:

- It factors out `scale` which could just be done more simply using `Mul.make_args(factor_terms(dif))`.
- It then expects to have an `Add` leftover and it will take some terms of the `Add` and move them to the other side. This just needs to use `Add.make_args` and check which terms don't involve the free symbols.

There is no reason that this should only work in the linear case and could not also apply to polynomials and I'm fairly sure that the code could be changed to do this more simply and handle both linear and polynomial inequalities.

Reply:
Thank you so much for such a detailed explanation of the situation and how I can make my way around it. I just pushed to my fork. The sun is about to rise where I am so I will check the CI next morning to see if everything passed.